### PR TITLE
A version two?

### PR DESCRIPTION
**Pull request by i509VCB**
Static data surprisingly still works on the latest version of fabric loader (0.9.0). Aside from that there are a few things static data does not cover well, and I have rewritten to in the end be more flexible.

This is sadly not backwards compatible.

Changes:
 - Addition of `DataSource`, an abstraction of the "source" of a static data item. A data source is either the "global" or from a mod (mod container).
 - All static data methods which used to return an `ImmutableSet` now just return `Set<StaticData.Item>`. The returned set is still unmodifiable (using Collections).
 - `StaticDataItem` is now a subclass of `StaticData`. `StaticDataItem` -> `StaticData.Item`
 - `StaticData.Item` is now an interface.
 - `StaticData` class is now final
 - Bump loom
 - Add testmod sourceset for some testing.

This is a draft PR since I will be writing a couple of tests first using the testmod source set.

Also I am going to hook up licenser and clean up the buildscript.

### CODE REVIEW

File: src/main/java/io/github/cottonmc/staticdata/DataSource.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+		@Override
+		Identifier createId(Path path) {
+			String rel = DataSource.getRelative(StaticData.GLOBAL_DATA_PATH, path);
+
+			if (rel.startsWith("static_data/")) { //Should always be true
+				rel = rel.substring("static_data/".length());
+			}
+
+			return new Identifier("global", rel);
```
Comment (LemmaEOF) [inline src/main/java/io/github/cottonmc/staticdata/DataSource.java:87]: should this remain as `g` like in version 1?

File: src/main/java/io/github/cottonmc/staticdata/ItemImpl.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+
+final class ItemImpl implements StaticData.Item {
```
Comment (LemmaEOF) [inline src/main/java/io/github/cottonmc/staticdata/ItemImpl.java:13]: Feels a bit weird to have this package-private but leave everything else public. Folks could implement their own `DataSource` and load static data from a new place. Probably not too much of a problem, honestly.

**Review by LemmaEOF — APPROVED** [review]

File: src/main/java/io/github/cottonmc/staticdata/ItemImpl.java (hunk 2 above)
Comment (Frontrider) [reply src/main/java/io/github/cottonmc/staticdata/ItemImpl.java:13]: As someone who was at the bottom of the idea:
The whole point of the mod is so you need not implement anything like that, and can use a data/resource pack like structure. 

The only stuff that is NOT covered right now and you may want to Implement it is namespaces in the external static data folder, and the usage of zip files as static data packs. If that is there then I honestly don't know what else we could require.

File: src/main/java/io/github/cottonmc/staticdata/DataSource.java (hunk 1 above)
Comment (Frontrider) [reply src/main/java/io/github/cottonmc/staticdata/DataSource.java:87]: Having actual namespacing in there may be a better solution, as that allows anyone to use that folder for much faster prototyping without having to find and replace literally every single ID that you'd use anywhere else.

From like: "setblock global:block" to "setblock modid:block"

**Comment by falkreon** [comment]
I feel like DataSource is either undesirable or misnamed.

A "data source" as stated is a String corresponding to a mod resource domain, or in the case of the server staticdata folder, "g" or "global".

Then there's a useful concept of "data providers", there being a distinct queryable list of objects that correspond to a particular resource-domain but can be individually queried for more data - probably as the result of information from a previous query.

I feel like the class suggests the latter but the name suggests the former.

I'd caution against forcing Identifiers used by the static-data system to take the same form as Mojang's resource system. You cannot necessarily ask Mojang's resource system to provide you with the data from that Identifier and get the same data, or any data at all. The "resource-domain" of a static-data identifier *must* correspond to the data-provider in question, otherwise there is literally no way to go back to that domain to look up additional/related data. Basically, the ability to lie undermines the whole system.

**Pull request closed without merging**